### PR TITLE
fix: CodeUri is not resolvable in nested stacks not located with root template

### PR DESCRIPTION
**Pull request by aahung**
#### Which issue(s) does this change fix?

#2692

#### Why is this change necessary?

Fix a bug causing relative resource CodeUri is not resolvable when the resources is defined in a nested stack whose template file is located in a different directory from root template.

For example, 

```
            root stack                -> template.yaml
            child stack               -> folder/template.yaml
            a resource in child stack -> folder/resource
```

The CodeUri of the resource is "./resource" (or without `./`). When sam tries to build it, it assume every codeuri is relative to the root template, which results in build failure. 

#### How does it address the issue?

When creating `Function` and `LayerVersion`, the `CodeUri` is replaced using `SamLocalStackProvider.normalize_resource_path()`.

This change applies to `Function` and `LayerVersion`'s `codeuri` and `Function`'s `metadata.DockerContext`.

Originally, in `local_labmda` and `layer_download` handles this problem themselves, now we have a universal solution.

#### Checklist

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [x] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/providers/sam_stack_provider.py [hunk 1]
```diff
@@ -290,3 +290,27 @@ def merge_parameter_overrides(
...
+        stack
+        path
+
+        Returns
+        -------
+        str
+            the normalized path relative to root stack
+
+        """
+        return path if os.path.isabs(path) else os.path.normpath(os.path.join(os.path.dirname(stack.location), path))
```
Comment (hoffa) [inline samcli/lib/providers/sam_stack_provider.py:316]: 1. A bit hard to follow with everything on one line.
2. Can we use `pathlib`? It should normalize paths by default which should simplify this.

File: samcli/lib/providers/sam_layer_provider.py [hunk 2]
```diff
@@ -105,8 +104,8 @@ def _convert_lambda_layer_resource(
 
         return LayerVersion(
             layer_logical_id,
-            codeuri,
+            codeuri and SamLocalStackProvider.normalize_resource_path(stack, codeuri),
```
Comment (hoffa) [inline samcli/lib/providers/sam_layer_provider.py:107]: Why is the `codeuri and ...` needed?

File: samcli/lib/providers/sam_stack_provider.py (hunk 1 above)
Comment (aahung) [reply samcli/lib/providers/sam_stack_provider.py:316]: `Path()` does not always normalize, for example, `Path('./a', '..')` would give `a/..` while `os.path.normpath(os.path.join('./a', '..'))` gives `.`

Comment (aahung) [reply samcli/lib/providers/sam_stack_provider.py]: made it 2 line!

File: samcli/lib/providers/sam_layer_provider.py (hunk 2 above)
Comment (aahung) [reply samcli/lib/providers/sam_layer_provider.py:107]: `codeuri` could be `None`, so I need to check it first.

Comment (hoffa) [reply samcli/lib/providers/sam_layer_provider.py:107]: Would it make sense to check that in `normalize_resource_path` instead?

Comment (aahung) [reply samcli/lib/providers/sam_layer_provider.py:107]: I made it consistent with built in path functions, such as `os.path.join(a, b)` or `Path(a, b)`, functions with "path" in their name do not accept `None` as an argument.

File: samcli/lib/providers/sam_function_provider.py [hunk 3]
```diff
@@ -277,32 +288,35 @@ def _build_function_configuration(
...
+            stack_path=stack.stack_path,
             name=name,
             functionname=resource_properties.get("FunctionName", name),
             packagetype=resource_properties.get("PackageType", ZIP),
             runtime=resource_properties.get("Runtime"),
             memory=resource_properties.get("MemorySize"),
             timeout=resource_properties.get("Timeout"),
             handler=resource_properties.get("Handler"),
-            codeuri=codeuri,
+            codeuri=codeuri and SamLocalStackProvider.normalize_resource_path(stack, codeuri),
```
Comment (sriram-mv) [inline samcli/lib/providers/sam_function_provider.py:304]:
```
SamLocalStackProvider.normalize_resource_path(stack, codeuri) if codeuri else None
```

would that be more descriptive? is that what you are going for?

File: samcli/lib/providers/sam_function_provider.py [hunk 4]
```diff
@@ -374,10 +388,10 @@ def _parse_layer_info(
                 layers.append(
                     LayerVersion(
                         layer_logical_id,
-                        codeuri,
+                        codeuri and SamLocalStackProvider.normalize_resource_path(stack, codeuri),
```
Comment (sriram-mv) [inline samcli/lib/providers/sam_function_provider.py:391]: same here

**Review by sriram-mv — CHANGES_REQUESTED** [review]

File: samcli/lib/providers/sam_function_provider.py (hunk 3 above)
Comment (aahung) [reply samcli/lib/providers/sam_function_provider.py]: ok, makes sense, consistent with `imageuri=imageuri if imageuri else resource_properties.get("ImageUri"),`

**Review by hoffa — APPROVED** [review]

**Comment by kornicameister** [comment]
hello everyone,
I am playing around with this PR because it will make some (or rather a lot of things) simpler on my end. Works like a charm apart from one thing. The way our templates are organized is via symlinks resembling some like

```
stack_1
stack_2
stack_3
  stacks:
    - stack_1
    - stack_2
```

Reason for that is that we can reference supplemental/extra sources with the same S3 object prefix and reuse those in templates. However even if I can do `Location: ../stack_1/template.yml` I cannot do `Location: ./stack_1/template.yml` (i.e. symlink).  

Perhaps this PR could be enhanced with resolving symlinks?

**Comment by aahung** [comment]
> hello everyone,
> I am playing around with this PR because it will make some (or rather a lot of things) simpler on my end. Works like a charm apart from one thing. The way our templates are organized is via symlinks resembling some like
> 
> ```
> stack_1
> stack_2
> stack_3
>   stacks:
>     - stack_1
>     - stack_2
> ```
> 
> Reason for that is that we can reference supplemental/extra sources with the same S3 object prefix and reuse those in templates. However even if I can do `Location: ../stack_1/template.yml` I cannot do `Location: ./stack_1/template.yml` (i.e. symlink).
> 
> Perhaps this PR could be enhanced with resolving symlinks?

Thanks for reporting, it will be fixed by https://github.com/aws/aws-sam-cli/pull/2697

File: tests/integration/buildcmd/test_build_cmd.py [hunk 5]
```diff
@@ -1585,8 +1588,59 @@ def test_nested_build(self, use_container, cached, parallel):
...
+            - FunctionB
+            - ChildStackY
+                - template.yaml
+                - FunctionA
+                - MyLayerVersion
+    """
+
+    template = os.path.join("deep-nested", "template.yaml")
+
+    # @pytest.mark.flaky(reruns=3)
```
Comment (sriram-mv) [inline tests/integration/buildcmd/test_build_cmd.py:1608]: was this test supposed to be flaky? if not lets remove this comment.

**Review by sriram-mv — APPROVED** [review]
minor nit: otherwise looks good.

File: tests/integration/buildcmd/test_build_cmd.py (hunk 5 above)
Comment (aahung) [reply tests/integration/buildcmd/test_build_cmd.py:1608]: good catch! I commented out it for local testing purpose, let me bring it back! thx

**Comment by aahung** [comment]
Accidentally put a commit in #2697 here, just reverted.

**Comment by aahung** [comment]
@kornicameister update: the symbolic link is supported in #2697, will be out in the next release.

**Comment by kornicameister** [comment]
Awesome. Thanks a lot.
